rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
+    ($data_ty:path, $native_ty:ident) => {
+        impl PrimitiveArray<$native_ty> {
+            pub fn new(length: i64, values: Buffer, null_count: i64, offset: i64) -> Self {
+                let array_data = ArrayData::builder($data_ty)
+                    .length(length)
+                    .add_buffer(values)
+                    .null_count(null_count)
+                    .offset(offset)
+                    .build();
+                PrimitiveArray::from(array_data)
+            }
+
+            /// Returns a `Buffer` holds all the values of this array.
+            pub fn values(&self) -> Buffer {
+                self.data.buffers()[0].copy()
+            }
+
+            /// Returns a raw pointer to the values of this array.
+            pub fn raw_values(&self) -> *const $native_ty {
+                unsafe { mem::transmute(self.raw_values.get().offset(self.data.offset() as isize)) }
+            }
+
+            /// Returns the primitive value at index `i`.
+            pub fn value(&self, i: i64) -> $native_ty {
+                unsafe { *(self.raw_values().offset(i as isize)) }
+            }
+
+            /// Determine the minimum value in the array
+            pub fn min(&self) -> Option<$native_ty> {
+                self.search(|a, b| a < b)
+            }
+
+            /// Determine the maximum value in the array
+            pub fn max(&self) -> Option<$native_ty> {
+                self.search(|a, b| a > b)
+            }
+
+            fn search<F>(&self, cmp: F) -> Option<$native_ty>
+            where
+                F: Fn($native_ty, $native_ty) -> bool,
+            {
+                let mut n: Option<$native_ty> = None;
+                let data = self.data();
+                for i in 0..data.length() {
+                    if data.is_null(i) {
+                        continue;
+                    }
+                    let mut m = self.value(i as i64);
+                    match n {
+                        None => n = Some(m),
+                        Some(nn) => if cmp(m, nn) {
+                            n = Some(m)
+                        },
+                    }
+                }
+                n
+            }
         }
-        ListArray::from(builder.finish())
-    }
+
+        /// Construct a primitive array from a vector. Should only be used for testing.
+        impl From<Vec<$native_ty>> for PrimitiveArray<$native_ty> {
+            fn from(data: Vec<$native_ty>) -> Self {
+                let array_data = ArrayData::builder($data_ty)
+                    .length(data.len() as i64)
+                    .add_buffer(Buffer::from(data.to_byte_slice()))
+                    .build();
+                PrimitiveArray::from(array_data)
+            }
+        }
+
+        impl From<Vec<Option<$native_ty>>> for PrimitiveArray<$native_ty> {
+            fn from(data: Vec<Option<$native_ty>>) -> Self {
+                const TY_SIZE: usize = mem::size_of::<$native_ty>();
```

Comment:
Wouldn't it be simpler to implement this using a `Builder<$native_ty>`?

Reply:
The current `Builder` only builds `Buffer` though and it doesn't handle null values. I do intend to replace this in future when the aforementioned `ArrayBuilder` is ready (similar to the c++ [`ArrayBuilder`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/builder.h#L228))

Reply:
Oh, right, I missed that. Sorry.